Treat macros the same as functions in docsystem

MichaelHatherly opened the pull request:
Since #14563 made macros into generic functions we should naturally handle them in the same way as functions in the docsystem.

Since the expression `@time` is no different to `@time()` this is slightly problematic for querying the docsystem for the docs associated with the `Function` rather than any one particular `Method` of the
macro. Hence I've added the syntax `help?> :@time` to get docs for the whole macro while `help?> @time`, `help?> @time(x)`, etc. will get docs for a particular signature. The quote (`:`) syntax mirrors that of

```
"..."
:@time
```

syntax which is already available for use to document `@time` so using it for retrieving as well as setting docs doesn't seem too much of a stretch. Any other suggestions on how to handle this better would be great.

I've also reorganised the `docm` method to be a bit less complex and added some comments describing the different syntax each case handles.

MichaelHatherly committed:
Treat macros the same as functions in docsystem

Macros are now treated as generic functions so we should naturally
handle them in the same way as functions in the docsystem.

Since the expression `@time` is no different to `@time()` this is
slightly problematic for querying the docsystem for the docs associated
with the `Function` rather than any one particular `Method` of the
macro. Hence I've added the syntax `help?> :@time` to get docs for the
whole macro while `help?> @time`, `help?> @time(x)`, etc. will get docs
for a particular signature. The quote (`:`) syntax mirrors that of

    "..."
    :@time

syntax which is already available for use to document `@time` so using
it for retrieving as well as setting docs doesn't seem too much of a
stretch.

tkelman commented:
Seems totally reasonable. :100: for comments

MichaelHatherly commented:
Currently this lacks some docs updates. Those depend on the syntax decision needed for macro lookup, i.e. `:@time` vs `@time`. I'll add the necessary docs changes once that's been settled.

nalimilan commented:
Couldn't you show all signatures when calling `help?> @time`? That would be only slightly annoying (too verbose) when you wanted that particular signature, but in most cases this would be much more natural than `help?> :@time` (which people won't probably try).

MichaelHatherly commented:
> Couldn't you show all signatures when calling `help?> @time`

Yeah, could do it that way. The number of 0-arg macros is quite low, `@__FILE__` come to mind as one exception, so it wouldn't have much of an impact I'd think.

Other option might be to do a tiny bit of pre-processing in https://github.com/JuliaLang/julia/blob/c1faf7fabf4d6184564ec0621f1b7fb51987b77a/base/REPL.jl#L717-L720 to differentiate between `@time` and `@time()` and handle each appropriately. That would match up better with how function call syntax in `help?>` mode is handled.

MichaelHatherly committed:
Improve REPL help when searching for macro docs

Distingush between bare macro names and the 0-arg macrocall to the
macro, i.e. `@macroname` and `@macroname()`. Also moves the bits of
helpmode handling that were in REPL.jl to the Docs module so it's
easier to find.

MichaelHatherly committed:
Update manual section on documenting macros

MichaelHatherly commented:
Should be good to go now I think. `help?> @macroname` returns docs for all signatures of `@macroname`, while `help?> @macroname()` returns those just for the 0-arg method if it exists. This matches up pretty well with how function docs are retrieved, so hopefully shouldn't be a surprising behaviour.

omus commented:
This reminds me that the documentation for macros without parenthesis don't get parsed correctly: https://github.com/JuliaLang/JuliaDoc/pull/18

MichaelHatherly commented:
Any further comments on this, else I'd like to merge this within the next few days so I can continue with a few other refactorings.

@omus's regex in https://github.com/JuliaLang/JuliaDoc/pull/18 looks like good to me, any reason not to merge that? (other than it slipping through the radar)

tkelman commented on test/docs.jl:
```diff
@@ -349,12 +384,7 @@ end
-
-    `@m2_11993` is a macro.""")
-end
+@test (@doc @m2_11993) === nothing
```
does this no longer give a message of any kind?

MichaelHatherly replied:
Yes, it now acts the same as searching for method docs when the signature doesn't match, returning `nothing`. I think I might be able to get the message back though, I'll have a look tomorrow.

MichaelHatherly replied:
Fixed up that case to correctly return the `No documentation found.` message now.

MichaelHatherly commented:
With @tkelman's comment addressed I'll merge this later this evening.
